pkg/kapp/cmd/app/deploy.go

```diff
@@ -220,18 +222,7 @@ func (o *DeployOptions) Run() error {
 	}
 
 	err = touch.Do(func() error {
-		defer func() {
-			if o.DeployFlags.AppMetadataFile != "" {
-				meta, err := app.Meta()
-				if err != nil {
-					return
-				}
-				err = os.WriteFile(o.DeployFlags.AppMetadataFile, []byte(meta.AsString()), os.ModePerm)
-				if err != nil {
-					return
-				}
-			}
-		}()
+		defer o.writeAppMetadataToFile(app)
```

Comment:
Is there any harm in doing this just once when we are sure we have the app (we can defer the function once, around line 116 I believe would work)?

I do not see why we *need* to write the file before `touch.Do` exits 🤔

Reply:
> Is there any harm in doing this just once when we are sure we have the app (we can defer the function once, around line 116 I believe would work)?

Good point!! I will check if that's doable.

> I do not see why we need to write the file before touch.Do exits 🤔

We need to save the metadata file even if there's no change to the app (we are returning early in that case and do not reach touch.Do)

Reply:
> We need to save the metadata file even if there's no change to the app (we are returning early in that case and do not reach touch.Do)

Yep that makes sense. I was wondering if the deferring needs to be done inside the function passed to `touch.Do` for some reason instead of doing it once earlier, that is all. I could not think of anything right off the bat.

Reply:
So, we have the [AskForConformation](https://github.com/vmware-tanzu/carvel-kapp/blob/68322f5ba9dc153aac6b76d5ee97255883a7a46f/pkg/kapp/cmd/app/deploy.go#L190) on line 190 and I think that we shouldn't write to metadata before this, i.e. before the user gives the confirmation. The only special case to this would be when there are no changes or diff run flag is being used (which is where I have added the defer call). I am still thinking what would be the right place to have a single defer call, but do let me know if you have any thoughts on this.

Reply:
Oh I see, so we only wanna write the file if
1. There are no changes
2. Changes get applied

Reply:
The only change I see in that case is that when there are no changes, the call does not need to be deferred.

Another benefit of this would be that when we grab new GKs before confirmation, those do not get written unless the changes have been applied.

Resolving this and putting a comment above.